src/promql/src/extension_plan/scalar_calculate.rs

```diff
@@ -0,0 +1,447 @@
+    }
+
+    fn from_template(&self, expr: &[Expr], inputs: &[LogicalPlan]) -> Self {
+        assert!(!inputs.is_empty());
+        ScalarCalculate {
+            schema: self.schema.clone(),
+            input: inputs[0].clone(),
+            project_index: self.project_index,
+            tag_columns: self.tag_columns.clone(),
+            exprs: expr.to_vec(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ScalarCalculateExec {
+    schema: SchemaRef,
+    project_index: (usize, usize),
+    input: Arc<dyn ExecutionPlan>,
+    tag_columns: Vec<String>,
+    metric: ExecutionPlanMetricsSet,
+}
+
+impl ExecutionPlan for ScalarCalculateExec {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        Partitioning::UnknownPartitioning(1)
+    }
+
+    fn required_input_distribution(&self) -> Vec<Distribution> {
+        vec![Distribution::SinglePartition]
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        self.input.output_ordering()
+    }
+
+    fn maintains_input_order(&self) -> Vec<bool> {
+        vec![true; self.children().len()]
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> DataFusionResult<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(ScalarCalculateExec {
+            schema: self.schema.clone(),
+            project_index: self.project_index,
+            tag_columns: self.tag_columns.clone(),
+            input: children[0].clone(),
+            metric: self.metric.clone(),
+        }))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> DataFusionResult<SendableRecordBatchStream> {
+        let baseline_metric = BaselineMetrics::new(&self.metric, partition);
+        let input = self.input.execute(partition, context)?;
+        let schema = input.schema();
+        let tag_indices = self
+            .tag_columns
+            .iter()
+            .map(|tag| {
+                schema
+                    .column_with_name(tag)
+                    .unwrap_or_else(|| panic!("tag column not found {tag}"))
+                    .0
+            })
+            .collect();
+        Ok(Box::pin(ScalarCalculateStream {
+            schema: self.schema.clone(),
+            project_index: self.project_index,
+            metric: baseline_metric,
+            tag_indices,
+            input,
+            have_multi_series: false,
+            done: false,
+            batch: None,
+            tag_value: None,
+        }))
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metric.clone_inner())
+    }
+
+    fn statistics(&self) -> Statistics {
+        Statistics {
```

Comment:
Not sure if this function's result is correct, @waynexia has to take a look.

Reply:
Looks like it's better to return the `self.input.statistics()`.

Reply:
Use `self.input.statistics()` may not a good option. In newer datafusion, it may lead crash because the change of schema.  this part still need to change when datafusion update is ready

Reply:
Got it. So maybe we can leave a `FIXME` here?

Reply:
The Statistics has fixed